Introduce OV Extension base api

ilyachur opened the pull request:
### Details:
 - Introduced base `ov::Extension` API
     - `ov::OpExtension` is needed to register custom operations
 - Implemented new template extension library
     - Library contains custom `Identity` operation
 - Extended `ov::runtime::Core` API
     - `add_extension("path_to_library")`
     - `add_extension(std::vector<ov::Extension::Ptr>&)`
     - `add_extension<Extension type, ...>(ext, ...)`
     - `add_extension<Operation op, ...>()`
 - New API supports old and new extensions to read models with custom operations
 - Added `add_extensions(std::vector<ov::Extension::Ptr>&)` to `Frontend` API
 - Changed `IR Frontend`:
    - Support of new extensions
    - removed `use_framework_node`
- Developed new tests to check that network with custom operation can be read with old and new extensions in new API 

### Tickets:
 - 58738

ilyachur commented:
The replacement for: https://github.com/openvinotoolkit/openvino/pull/7204

apankratovantonp commented on ngraph/core/include/openvino/core/extension.hpp:
```diff
@@ -0,0 +1,73 @@
+
+namespace ov {
+
+class OPENVINO_API Extension {
```
enable shared_from_this

ilyachur replied:
Why do we need to have shared_from_this for extensions?

apankratovantonp replied:
It is intended to be used just with `std::shared_ptr` is not it? So it is just trivial optimization to place reference_cointer inside Extension class and reduce heap fragmentation a little bit in cases when it is created not by std::make_shared.

apankratovantonp commented on ngraph/core/include/openvino/core/extension.hpp:
```diff
@@ -0,0 +1,73 @@
+class OPENVINO_API Extension {
+public:
+    using Ptr = std::shared_ptr<Extension>;
+    virtual ~Extension();
```
make destructor protected?

apankratovantonp commented on docs/template_extension/new/op.cpp:
```diff
@@ -0,0 +1,107 @@
+//! [op:ctor]
+
+//! [op:validate]
+void Operation::validate_and_infer_types() {
```
I think we need a more complex example operation to show how implement all overloaded methods

ilyachur replied:
It is a copy/past of existed operation

apankratovantonp commented on docs/template_extension/new/op.cpp:
```diff
@@ -0,0 +1,107 @@
+
+bool Operation::evaluate(const ngraph::HostTensorVector& outputs, const ngraph::HostTensorVector& inputs) const {
+    switch (inputs[0]->get_element_type()) {
+    case ngraph::element::Type_t::i8:
```
We need some helper function to reduce such `switch-case` bloat. For example in `arm-plugin` https://github.com/openvinotoolkit/openvino_contrib/blob/bd192a7a1cca7d580f5446dbb8473e2229807fff/modules/arm_plugin/src/arm_converter/arm_converter_gather.cpp#L43 
But it uses c++14

apankratovantonp commented:
Just want to add. The current solution is not only inconvenient for users but also UNSAFE. Due to exceptions, the `unload` function won't be called even by the user. I strongly recommend moving `so` management to extension containing classes before merging PR!

ilyachur commented:
> Just want to add. The current solution is not only inconvenient for users but also UNSAFE. Due to exceptions, the `unload` function won't be called even by the user. I strongly recommend moving `so` management to extension containing classes before merging PR!

In general user cannot get an access to extensions from shared library. Only our components work with these extensions.
Any way, I added internal SOExtension class to control the process of destroying extension objects.

ilyachur commented:
@apankratovantonp @ilya-lavrenov As I said, I changed the mechanism and added SOExtension.
It allows to avoid unload functions.
So I merge this PR.
If you have any other comments I will fix their in the separate PR.

ilya-lavrenov commented on ngraph/core/src/so_extension.hpp:
```diff
@@ -0,0 +1,49 @@
+private:
+    std::shared_ptr<void> m_so;
+    Extension::Ptr m_ext;
+};
```
can we just move `SoPtr` from IE to `ov::utils` and reuse here?
